README
                   --+-- -- -- -- -- -- --+--
                  +--       CTAFCONF       --+
                   --+-- -- -- -- -- -- --+--



===========================================================
INSTALL
===========================================================
the ctafconf should be installed in ~/.config/ctafconf.
Installation should be done using the git repository,
if you dont have git access you can get a tarball on github.
Always prefer git master, the branch is keept as stable as possible.
After calling ct-installconf, edit user-profile.sh to change your settings.


Prefered way to install, from GIT:
> mkdir .config
> cd .config
> git clone git://github.com/ctaf42/ctafconf.git
> cd ctafconf/bin
> ./ct-installconf
> nano ~/.config/ctafconf/user-profile.sh

Install from a tarball.
 you download the tarball from the website and unpack it in .config
> mkdir .config
> cd .config
> tar xvzf ctafconf-DATE.tgz
> cd ctafconf/bin
> ./ct-installconf
> nano ~/.config/ctafconf/user-profile.sh


===========================================================
UPDATE
===========================================================
To update your configuration just call git pull, or extract newer file over the previous one.
That's all, you dont need to call ct-installconf or ct-updateconf, it's automatic

Using git:
>cd .config/ctafconf
>git pull
That's all, the next time you launch your shell, the ctafconf will be updated

using a newer tarball:
unpack the tarball in .config
>cd .config/
>tar xvzf ctafconf-DATE.tgz


==========================================================
INSTALL FOR EMACS ONLY
==========================================================
if you are paranoid, and dont trust the install script.
Try the manual way to get the emacs configuration:
>echo "(load-file \"~/.config/ctafconf/etc/emacs/emacs\")" >> ~/.emacs


===========================================================
VIM
===========================================================
See ~/.config/ctafconf/etc/vim/README



===========================================================
ZSH
===========================================================
- ctrl-c in completion menu go back to the text previously enterred
- up/down search through history (start typing then up/down to see what i mean)
- pageup/pagedown cycle through history (it does not care about what is already on the command line)
- ctrl-x h call the man for the first word on the command line
- ctrl-x e edit the current command line in your editor
- ctrl-x w display the location of the current command
- ctrl-x c calculator

===========================================================
UNINSTALL
===========================================================
You will never want to uninstall, after you tried the ctafconf

===========================================================
HELP
===========================================================
It's a good framework to create our own personnal configuration

ct-ssh        : create alias for ssh host
findr         : recursive find
grepr         : recursive grep with color
rpurge        : cleanup temporary file (#*, *~)
rpurge_c      : cleanup c compiled object (*.o)
rpurge_cmake  : cleanup cmake generated makefiles (Makefile, CMakeCache.txt, ..)
rpurge_svn    : cleanup svn files (.svn)
rpurge_python : cleanup python files (*.pyc, *.pyo)
rpurge_cvs    : cleanup cvs files (.cvs)
replacer      : recursive find and replace in all files in the current folder

visit http://www.ctaf.free.fr for more information/help about the ctafconf

===========================================================
Credits
===========================================================
Some of thoses configurations files have been collected, modified
and enhanced from a lot's of files from differents places.
thanks for all peoples who share configs on the web.
--
thanks for Sergio Talens-Oliag <sto@debian.org> for his ssft library
thanks for all people I 'stolen' GPL code ;-)


===========================================================
About
===========================================================
Cedric GESTES (ctaf42@gmail.com)
http://www.sf.net/projects/grk
http://grk.sf.net
http://www.ctaf.free.fr

 _____________________________________
< CTAFCONF? yuhh? MeuuuuuuuuuHHHHHHHH >
 -------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


ChangeLog
2010-06-12  Cedric GESTES  <ctaf42@gmail.com>
	* ct-installconf: source could be outside ~/.config/ctafconf

2010-06-08  Cedric GESTES  <ctaf42@gmail.com> Massive Changelog Update
	* ctafconf: install rewriten in python
	* ctafconf: no more ~/.config/ctafconf/perso => personal files are next to config file, (eg ~/.zshrc.user for zsh)
	* ctafconf: support profiles (there is 3 profile atm: aldebaran,ctaf,yannick)
	* ctafconf: no more ct-profile, a simple config file should be tweaked in ~/.config/ctafconf/user-profile.sh
	* emacs: massive improvement/cleanup
	* emacs: include ecb (emacs code browser), pymacs
	* emacs: support completion whenever it's possible using company (press TAB to try)
	         (complete python code with ropemacs, complete c/c++ with semantics)
	* emacs: framed speedbar (Shift-F5)
	* emacs: support profile (only ctaf profile include emacs tweak)
	* emacs: removed each package already provided by emacs23.
	* emacs: provide a visual folding mode for elisp/python/c/c++ (just click in the fringe)
	* emacs: updated yasnippet
	* emacs: display a menu with function/variable/class for fast access (available for most programming language)
	* emacs: added haskell, protobuf maybe more

2010-01-09  Cedric GESTES  <ctaf42@gmail.com>
	* shell/alias: added sh specific alias
	* sz: find then source .env files. start from working directory descending to /

2009-09-12  Cedric GESTES  <ctaf42@gmail.com>
	* emacs/template: added .cmake template

2009-09-06  Cedric GESTES  <ctaf42@gmail.com>
	* etc/zsh/zshalias: add mmv => simple zmv

2009-08-19  Cedric GESTES  <ctaf42@gmail.com>
	* etc/zsh/zshstyle: added intuitive search in completion menu

2009-08-14  Cedric GESTES  <ctaf42@gmail.com>
	* zshoptions: command line correction desactivated (avoid bug with newly installed program)
	* zsh: ctrl-x c to get a calculator with history
	* zsh: always use menucomplete
	* zsh: ctrl-c in the completion menu go back before completion

2009-08-13  Cedric GESTES  <ctaf42@gmail.com>
	* prompt_ctaf_setup: refactored zsh compliant prompt
	* prompt_ctaf_setup: display info from VCS(git, svn, hg, ...)
	* zshrc: start using autoload
	* ct-updateconf: always check ctafconf git not the current folder
	* zsh: silent degraded mode when not all feature are found

2009-08-11  Cedric GESTES  <ctaf42@gmail.com>
	* bin/ct-updateconf: autoupdate based on git commit
	* zsh: splitted config files (options, style, keys)
	* zshprompt: correct exit status
	* etc/shell/alias: ka for killall
	* etc/shell/alias: 644, 755 for fast chmod
	* zsh: bigger history to avoid typing the same command again and again
	* zshalias: added ... .... and ..... alias for ../.. etc.
	* zsh: find new installed program instead of complaining that the binary doesnt not exist
	* zsh: ctrl-x e to edit the current command line in editor
	* zsh: ctrl-x w to call which
	* etc/shell/env: append color in man
	* zsh: added bindkey for pageup/pagedown => up/down in history
	* zsh: better completion (cursor move to the next non-ambiguous letter)
        * zsh: use cache completion for everything
	* zsh: complete man page section (ctrl-h)
	* zsh: a new shell get the history from all other shell still open!
	* zsh: dont accept a match if it is ambiguous

2009-07-28  Cedric GESTES  <ctaf42@gmail.com>
	* etc/emacs/templates/makefile.tpl: fix a problem with tabulation and space

2009-04-18  Cedric GESTES  <ctaf42@gmail.com>
	* bin/rpurge_*: better name and explicit message
	* bin/rpurge_python : remove python compiled scripts

2009-04-16  Cedric GESTES  <gestes@aldebaran-robotics.com>
	* etc/emacs/templates/makefile.tpl: support install

2009-03-20  Cedric GESTES  <ctaf42@gmail.com>
	* etc/emacs/bindings.emacs: inverted M-Up and M-Down

2009-04-04  Cedric GESTES  <ctaf42@gmail.com>
	* etc/emacs/emacs: added yasnippet

2009-03-18  Cedric GESTES  <ctaf42@gmail.com>
	* etc/emacs/templates : shebang for python template

2009-02-27  Cedric GESTES  <gestes@aldebaran-robotics.com>
	* etc/emacs/prog.emacs (err): added support for lua

2009-02-06  Cedric GESTES  <gestes@aldebaran-robotics.com>
	* etc/emacs/templates/makefile.tpl: more c++ friendly, added automatic dependencies

2009-01-26  Cedric GESTES  <ctaf42@gmail.com>
	* bin/colormake.pl: better color on template error

2009-01-25  Cedric GESTES  <ctaf42@gmail.com>
	* etc/zsh/zshconf : always expand completion if possible

2009-01-25  Cedric GESTES  <ctaf42@gmail.com>
	* etc/emacs/color.emacs : emacs 23 tweak

2009-01-22  Cedric GESTES  <ctaf42@gmail.com>
	* emacs: template for hpp

2009-01-05  Cedric GESTES  <ctaf42@gmail.com>
	* shell: max 1042 process (avoid forkbomb)
	* shell: added warning to indicate user which file they can edit

2008-10-07  Cedric GESTES  <ctaf42@gmail.com>
	* etc/shlib/xsession.sh (wmlist): dont consider terms like wm (in random mode) !!

	* etc/fluxbox/init: disable wallpaper (use the gnome one by default)

2008-10-06  Cedric GESTES  <ctaf42@gmail.com>
	* xsession : fixed: each call to test_wm return a random wm when wm is 'random'

2008-10-01  Cedric GESTES  <gestes@aldebaran-robotics.com>
	* etc/emacs/templates/makefile.tpl: support overriding CC, CXX, CFLAGS, LDFLAGS ! (for cross compil)

2008-09-30  Cedric GESTES  <ctaf42@gmail.com>

etc/awesome/rc.lua
-- standard function :
-- F2 : terminal
-- F5 : terminal
-- F2 : terminal
-- F2 : terminal


-- pomme+shift+left/right => change workspace
-- pomme+left/right => change window in group
-- pomme+up/down => change group




-- Include awesome libraries, with lots of useful function!
require("awful")
require("beautiful")

home = os.getenv("HOME")

-- {{{ Variable definitions
-- Themes define colours, icons, and wallpapers
-- The default is a dark theme
--theme_path = "/usr/share/awesome/themes/default/theme"
-- Uncommment this for a lighter theme
--theme_path = "/usr/share/awesome/themes/sky/theme"

theme_path = home .. "/.config/awesome/themes/default/theme"

-- Actually load theme
beautiful.init(theme_path)

-- This is used later as the default terminal and editor to run.
terminal = "gnome-terminal"
editor = os.getenv("EDITOR") or "nano"
editor_cmd = terminal .. " -e " .. editor

-- Default modkey.
-- Usually, Mod4 is the key with a logo between Control and Alt.
-- If you do not like this or do not have such a key,
-- I suggest you to remap Mod4 to another key using xmodmap or other tools.
-- However, you can use another modifier like Mod1, but it may interact with others.
modkey = "Mod4"

-- Table of layouts to cover with awful.layout.inc, order matters.
layouts =
{
    "tile",
    "tileleft",
    "tilebottom",
    "tiletop",
    "fairh",
    "fairv",
    "magnifier",
    "max",
    "fullscreen",
    "spiral",
    "dwindle",
    "floating"
}

-- Table of clients that should be set floating. The index may be either
-- the application class or instance. The instance is useful when running
-- a console app in a terminal like (Music on Console)
--    xterm -name mocp -e mocp
floatapps =
{
    ["MPlayer"] = true,
    ["Sonata"] = true,
    ["Pidgin"] = true,
    ["Gajim"] = true
}

-- Applications to be moved to a pre-defined tag by class or instance.
-- Use the screen and tags indices.
apptags =
{
    ["Emacs"] = { screen = 1, tag = 4 },
    ["Thunderbird"] = { screen = 1, tag = 3 },

    ["Navigator"] = { screen = 1, tag = 2 },
    ["Epiphany"] = { screen = 1, tag = 2 },

    ["Pidgin"]   =  { screen = 1, tag = 5 }
    --["Exaile.py"] = { screen = 1, tag = 3 },
    -- ["mocp"] = { screen = 2, tag = 4 },
}

-- Define if we want to use titlebar on all applications.
use_titlebar = false
-- }}}

-- {{{ Tags
-- Define tags table.
-- }}}
tags_name   = { "1:main", "2:www", "3:mail", "4:prog", "5:im", "6:float"}
tags_layout = { 1     , 8    , 8     ,  8    , 12  ,   12   }

tags = {}
for s = 1, screen.count() do
    -- Each screen has its own tag table.
    tags[s] = {}
  -- Create all tags on each screen
  for tagnumber, tagname in ipairs(tags_name) do
    tags[s][tagnumber] = tag({ name = tagname, layout = layouts[tags_layout[tagnumber]] })
    -- Add tags to screen one by one
    tags[s][tagnumber].screen = s
  end
    -- I'm sure you want to see at least one tag.
    tags[s][1].selected = true
end

-- {{{ Wibox
-- Create a textbox widget
mytextbox = widget({ type = "textbox", align = "right" })
-- Set the default text in textbox
--mytextbox.text = "<b><small> " .. AWESOME_RELEASE .. " </small></b>"

-- Create a laucher widget and a main menu
myawesomemenu = {
   { "manual", terminal .. " -e man awesome" },
   { "edit config", editor_cmd .. " " .. awful.util.getdir("config") .. "/rc.lua" },
   { "restart", awesome.restart },
   { "quit", awesome.quit }
}

mymainmenu = awful.menu.new({ items = { { "awesome", myawesomemenu, beautiful.awesome_icon },
                                        { "open terminal", terminal },
                                        { "thunderbird", "thunderbird3" } ,
                                        { "epiphany", "epiphany" } ,
                                        { "firefox", "firefox" } ,
                                        { "halt", "sudo /sbin/halt" },
                                        { "reboot", "sudo /sbin/reboot" }
                                      }
                            })

mylauncher = awful.widget.launcher({ image = beautiful.awesome_icon,
                                     menu = mymainmenu })

-- Create a systray
mysystray = widget({ type = "systray", align = "right" })

-- Create a wibox for each screen and add it
mywibox = {}
mypromptbox = {}
mylayoutbox = {}
mytaglist = {}
mytaglist.buttons = { button({ }, 1, awful.tag.viewonly),
                      button({ modkey }, 1, awful.client.movetotag),
                      button({ }, 3, function (tag) tag.selected = not tag.selected end),
                      button({ modkey }, 3, awful.client.toggletag),
                      button({ }, 4, awful.tag.viewnext),
                      button({ }, 5, awful.tag.viewprev) }
mytasklist = {}
mytasklist.buttons = { button({ }, 1, function (c) client.focus = c; c:raise() end),
                       button({ }, 3, function () awful.menu.clients({ width=250 }) end),
                       button({ }, 4, function () awful.client.focus.byidx(1) end),
                       button({ }, 5, function () awful.client.focus.byidx(-1) end) }

for s = 1, screen.count() do
    -- Create a promptbox for each screen
    mypromptbox[s] = widget({ type = "textbox", align = "left" })
    -- Create an imagebox widget which will contains an icon indicating which layout we're using.
    -- We need one layoutbox per screen.
    mylayoutbox[s] = widget({ type = "imagebox", align = "right" })
    mylayoutbox[s]:buttons({ button({ }, 1, function () awful.layout.inc(layouts, 1) end),
                             button({ }, 3, function () awful.layout.inc(layouts, -1) end),
                             button({ }, 4, function () awful.layout.inc(layouts, 1) end),
                             button({ }, 5, function () awful.layout.inc(layouts, -1) end) })
    -- Create a taglist widget
    mytaglist[s] = awful.widget.taglist.new(s, awful.widget.taglist.label.all, mytaglist.buttons)

    -- Create a tasklist widget
    mytasklist[s] = awful.widget.tasklist.new(function(c)
                                                  return awful.widget.tasklist.label.currenttags(c, s)
                                              end, mytasklist.buttons)

    -- Create the wibox
    mywibox[s] = wibox({ position = "top", fg = beautiful.fg_normal, bg = beautiful.bg_normal })
    -- Add widgets to the wibox - order matters
    mywibox[s].widgets = { mylauncher,
                           mytaglist[s],
                           mytasklist[s],
                           mypromptbox[s],
                           mytextbox,
                           mylayoutbox[s],
                           s == 1 and mysystray or nil }
    mywibox[s].screen = s
end
-- }}}

-- {{{ Mouse bindings
awesome.buttons({
    button({ }, 3, function () mymainmenu:toggle() end),
    button({ }, 4, awful.tag.viewnext),
    button({ }, 5, awful.tag.viewprev)
})
-- }}}

-- {{{ Key bindings

-- Bind keyboard digits
-- Compute the maximum number of digit we need, limited to 9
keynumber = 0
for s = 1, screen.count() do
   keynumber = math.min(9, math.max(#tags[s], keynumber));
end

tagkeys = { '#10', '#11', '#12', '#13', '#14', '#15', '#16', '#17', '#18', '#19', '#20' }
for i = 1, keynumber do
   keybinding({ modkey }, tagkeys[i],
              function ()
                 local screen = mouse.screen
                 if tags[screen][i] then
                    awful.tag.viewonly(tags[screen][i])
                 end
              end):add()
   keybinding({ modkey, "Control" }, tagkeys[i],
              function ()
                 local screen = mouse.screen
                 if tags[sCreen][i] then
                    tags[screen][i].selected = not tags[screen][i].selected
                 end
              end):add()
   keybinding({ modkey, "Shift" }, tagkeys[i],
              function ()
                 if client.focus then
                    if tags[client.focus.screen][i] then
                       awful.client.movetotag(tags[client.focus.screen][i])
                    end
                 end
              end):add()
   keybinding({ modkey, "Control", "Shift" }, tagkeys[i] ,
              function ()
                 if client.focus then
                    if tags[client.focus.screen][i] then
                       awful.client.toggletag(tags[client.focus.screen][i])
                    end
                 end
              end):add()
end

keybinding({ modkey }, "Escape", awful.tag.history.restore):add()

-- Standard program
keybinding({ modkey }, "Return", function () awful.util.spawn(terminal) end):add()

keybinding({ modkey }, "F2", function() awful.util.spawn(terminal) end):add()
keybinding({ modkey }, "F5", function() awful.util.spawn("nautilus --no-desktop") end):add()
keybinding({ modkey }, "F6", function() awful.util.spawn("epiphany") end):add()
keybinding({ modkey }, "F7", function() awful.util.spawn("thunderbird3") end):add()
keybinding({ modkey }, "F8", function() awful.util.spawn("totem") end):add()

--keybinding({ modkey, }, "x", function() awful.util.spawn("xlock") end):add()

keybinding({ modkey, "Control" }, "r", function ()
                                           mypromptbox[mouse.screen].text =
                                               awful.util.escape(awful.util.restart())
                                        end):add()
keybinding({ modkey, "Shift" }, "q", awesome.quit):add()

-- Client manipulation
keybinding({ modkey }, "m", awful.client.maximize):add()
keybinding({ modkey, "Shift" }, "Return", function () if client.focus then client.focus.fullscreen = not client.focus.fullscreen end end):add()

keybinding({ modkey }, "c", function () if client.focus then client.focus:kill() end end):add()
--keybinding({ "Alt" }, "c", function () if client.focus then client.focus:kill() end end):add()


keybinding({ modkey, "Shift" }, "Left",
           function ()
              if client.focus then
                 awful.tag.viewidx(-1, client.focus.screen)
              else
                 awful.tag.viewidx(-1)
              end
           end):add()

keybinding({ modkey, "Shift" }, "Right",
           function ()
              if client.focus then
                 awful.tag.viewidx(1, client.focus.screen)
              else
                 awful.tag.viewidx(1)
              end
           end):add()


local capi =
{
    screen = screen,
    client = client
}

--- Give the focus to a screen, and move pointer.
-- @param i Relative screen number.
function nomousefocus(i)
    local s = awful.util.cycle(capi.screen.count(), i)
    local c = awful.client.focus.history.get(s, 0)
    if c then capi.client.focus = c end

    -- dont move the mouse => xinemara fail otherwise...
    -- Move the mouse on the screen
    -- capi.mouse.screen = s
end


-- keybinding({ modkey }, "Up", function () focus(1) end):add()
-- keybinding({ modkey }, "Down", function () focus(2) end):add()
-- keybinding({ modkey }, "Up", function () awful.screen.focus(1) end):add()
-- keybinding({ modkey }, "Down", function () awful.screen.focus(2) end):add()
keybinding({ modkey }, "Up", function () nomousefocus(1) end):add()
keybinding({ modkey }, "Down", function () nomousefocus(2) end):add()

keybinding({ modkey }, "Left", function () awful.client.focus.byidx(-1); if client.focus then client.focus:raise() end end):add()
keybinding({ modkey }, "Right", function () awful.client.focus.byidx(1);  if client.focus then client.focus:raise() end end):add()


--keybinding({ modkey, "Shift" }, "Left", function () awful.client.focus.bydirection("left"); if client.focus then client.focus:raise() end end):add()
--keybinding({ modkey, "Shift" }, "Right", function () awful.client.focus.bydirection("right");  if client.focus then client.focus:raise() end end):add()
--keybinding({ modkey, "Shift" }, "Up", function () awful.client.focus.bydirection("up"); if client.focus then client.focus:raise() end end):add()
--keybinding({ modkey, "Shift" }, "Down", function () awful.client.focus.bydirection("down");  if client.focus then client.focus:raise() end end):add()

keybinding({ modkey, "Shift" }, "j", function () awful.client.swap.byidx(1) end):add()
keybinding({ modkey, "Shift" }, "k", function () awful.client.swap.byidx(-1) end):add()
keybinding({ modkey, "Control" }, "j", function () awful.screen.focus(1) end):add()
keybinding({ modkey, "Control" }, "k", function () awful.screen.focus(-1) end):add()

keybinding({ modkey, "Control" }, "space", awful.client.togglefloating):add()
keybinding({ modkey, "Control" }, "Return", function () if client.focus then client.focus:swap(awful.client.getmaster()) end end):add()
keybinding({ modkey }, "o", awful.client.movetoscreen):add()
keybinding({ modkey }, "Tab", awful.client.focus.history.previous):add()
keybinding({ modkey }, "u", awful.client.urgent.jumpto):add()
keybinding({ modkey, "Shift" }, "r", function () if client.focus then client.focus:redraw() end end):add()

-- Layout manipulation
keybinding({ modkey }, "l", function () awful.tag.incmwfact(0.05) end):add()
keybinding({ modkey }, "h", function () awful.tag.incmwfact(-0.05) end):add()
keybinding({ modkey, "Shift" }, "h", function () awful.tag.incnmaster(1) end):add()
keybinding({ modkey, "Shift" }, "l", function () awful.tag.incnmaster(-1) end):add()
keybinding({ modkey, "Control" }, "h", function () awful.tag.incncol(1) end):add()
keybinding({ modkey, "Control" }, "l", function () awful.tag.incncol(-1) end):add()
keybinding({ modkey }, "space", function () awful.layout.inc(layouts, 1) end):add()
keybinding({ modkey, "Shift" }, "space", function () awful.layout.inc(layouts, -1) end):add()

-- Prompt
keybinding({ modkey }, "F3", function ()
                                 awful.prompt.run({ prompt = "Run: " }, mypromptbox[mouse.screen], awful.util.spawn, awful.completion.bash,
                                                  awful.util.getdir("cache") .. "/history")
                             end):add()
keybinding({ modkey }, "F12", function ()
                                 awful.prompt.run({ prompt = "Run Lua code: " }, mypromptbox[mouse.screen], awful.util.eval, awful.prompt.bash,
                                                  awful.util.getdir("cache") .. "/history_eval")
                             end):add()

keybinding({ modkey, "Ctrl" }, "i", function ()
                                        local s = mouse.screen
                                        if mypromptbox[s].text then
                                            mypromptbox[s].text = nil
                                        elseif client.focus then
                                            mypromptbox[s].text = nil
                                            if client.focus.class then
                                                mypromptbox[s].text = "Class: " .. client.focus.class .. " "
                                            end
                                            if client.focus.instance then
                                                mypromptbox[s].text = mypromptbox[s].text .. "Instance: ".. client.focus.instance .. " "
                                            end
                                            if client.focus.role then
                                                mypromptbox[s].text = mypromptbox[s].text .. "Role: ".. client.focus.role
                                            end
                                        end
                                    end):add()

-- Client awful tagging: this is useful to tag some clients and then do stuff like move to tag on them
keybinding({ modkey }, "t", awful.client.togglemarked):add()

for i = 1, keynumber do
    keybinding({ modkey, "Shift" }, "F" .. i,
                   function ()
                       local screen = mouse.screen
                       if tags[screen][i] then
                           for k, c in pairs(awful.client.getmarked()) do
                               awful.client.movetotag(tags[screen][i], c)
                           end
                       end
                   end):add()
end
-- }}}

-- {{{ Hooks
-- Hook function to execute when focusing a client.
awful.hooks.focus.register(function (c)
    if not awful.client.ismarked(c) then
        c.border_color = beautiful.border_focus
    end
end)

-- Hook function to execute when unfocusing a client.
awful.hooks.unfocus.register(function (c)
    if not awful.client.ismarked(c) then
        c.border_color = beautiful.border_normal
    end
end)

-- Hook function to execute when marking a client
awful.hooks.marked.register(function (c)
    c.border_color = beautiful.border_marked
end)

-- Hook function to execute when unmarking a client.
awful.hooks.unmarked.register(function (c)
    c.border_color = beautiful.border_focus
end)

-- Hook function to execute when the mouse enters a client.
awful.hooks.mouse_enter.register(function (c)
    -- Sloppy focus, but disabled for magnifier layout
    if awful.layout.get(c.screen) ~= "magnifier"
        and awful.client.focus.filter(c) then
        client.focus = c
    end
end)

-- Hook function to execute when a new client appears.
awful.hooks.manage.register(function (c)
    if use_titlebar then
        -- Add a titlebar
        awful.titlebar.add(c, { modkey = modkey })
    end
    -- Add mouse bindings
    c:buttons({
        button({ }, 1, function (c) client.focus = c; c:raise() end),
        button({ modkey }, 1, function (c) c:mouse_move() end),
        button({ modkey }, 3, function (c) c:mouse_resize() end)
    })
    -- New client may not receive focus
    -- if they're not focusable, so set border anyway.
    c.border_width = beautiful.border_width
    c.border_color = beautiful.border_normal

    -- Check if the application should be floating.
    local cls = c.class
    local inst = c.instance
    if floatapps[cls] then
        c.floating = floatapps[cls]
    elseif floatapps[inst] then
        c.floating = floatapps[inst]
    end

    -- Check application->screen/tag mappings.
    local target
    if apptags[cls] then
        target = apptags[cls]
    elseif apptags[inst] then
        target = apptags[inst]
    end
    if target then
        c.screen = target.screen
        awful.client.movetotag(tags[target.screen][target.tag], c)
    end

    -- Do this after tag mapping, so you don't see it on the wrong tag for a split second.
    client.focus = c

    -- Set the windows at the slave,
    -- i.e. put it at the end of others instead of setting it master.
    -- awful.client.setslave(c)

    -- Honor size hints: if you want to drop the gaps between windows, set this to false.
    c.honorsizehints = false
end)

-- Hook function to execute when arranging the screen.
-- (tag switch, new client, etc)
awful.hooks.arrange.register(function (screen)
    local layout = awful.layout.get(screen)
    if layout then
        mylayoutbox[screen].image = image(beautiful["layout_" .. layout])
    else
        mylayoutbox[screen].image = nil
    end

    -- Give focus to the latest client in history if no window has focus
    -- or if the current window is a desktop or a dock one.
    if not client.focus then
        local c = awful.client.focus.history.get(screen, 0)
        if c then client.focus = c end
    end

    -- Uncomment if you want mouse warping
--     [[
--     if client.focus then
--         local c_c = client.focus:fullgeometry()
--         local m_c = mouse.coords()

--         if m_c.x < c_c.x or m_c.x >= c_c.x + c_c.width or
--             m_c.y < c_c.y or m_c.y >= c_c.y + c_c.height then
--             if table.maxn(m_c.buttons) == 0 then
--                 mouse.coords({ x = c_c.x + 5, y = c_c.y + 5})
--             end
--         end
--     end
--     ]]
end)

-- Hook called every second
awful.hooks.timer.register(1, function ()
--    mem = awful.util.pread(home .. "/bin/mem.py")
--    cpu = awful.util.pread(home .. "/bin/cpu.py")
   mem = "42"
   cpu = "42"
   date = os.date()
    mytextbox.text = " | " .. "Mem: " .. mem .. ", Cpu: " .. cpu .. " | " .. date .. " | " end)
-- }}}

-- Load revelation
require("revelation")

keybinding({ modkey }, "F1", revelation.revelation):add()
